[MkFit] Fix UBSAN error - skip processing of layers with no hits.

osschar opened the pull request:
When processing layers with no hits the LayerOfHits vector was in default state with storage pointer set to nullptr.

Loaders of hits for vector processing were still being initialized, resulting in doing pointer arithmetic with nullptr -- though they were never actually used.

This should fix #37136. Note that the wf there was 10004.0, single photon 10GeV and there were indeed several occurrences of empty mkfit hit containers.

osschar committed:
[MkFit] Fix UBSAN error - skip processing of layers with no hits.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37146/28688

- This PR adds an extra 24KB to repository

cmsbuild commented:
A new Pull Request was created by  @osschar (Matevž Tadel) for master.

It involves the following packages:

- RecoTracker/MkFitCore (**reconstruction**)


@jpata, @cmsbuild, @clacaputo, @slava77 can you please review it and eventually sign? Thanks.
@makortel, @felicepantaleo, @GiacomoSguazzoni, @JanFSchulte, @rovere, @VinInn, @ebrondol, @gpetruc, @mmusich, @mtosi, @dgulhan this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
test parameters:
  - release = CMSSW_12_3_UBSAN_X
  - workflow = 10004.0

makortel commented:
@cmsbuild, please test

mmusich commented:
@smuzaffar this has been running tests for 3 days (!?).
Is there something stuck?

smuzaffar commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cdca02/22911/summary.html
**COMMIT**: 81e7728c2b39c5a608e7949dc3ff80e79f14d0b7
**CMSSW**: CMSSW_12_3_X_2022-03-07-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/37146/22911/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-cdca02/10004.0_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 27 differences found in the comparisons
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3984447
 - DQMHistoTests: Total failures: 864
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3983561
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 204 log files, 45 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2022-03-07-1100+cdca02/48775/triggerResults): no differences found

clacaputo commented:
> @mmusich , UBSAN IB comparison does not work. The job timedout after 16 hours ( https://cmssdt.cern.ch/jenkins/job/compare-root-files-short-matrix/48727/ ) . Check the relvals for wf 10004.0 to see if USBAN issue is fixed or not?
> 
> I have restarted the comparison job now but I am afraid it will timeout again.

Hi @smuzaffar, where can I check wf 10004.0 relvals for UBSAN?

smuzaffar commented:
URL (i.e. https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cdca02/22856)  for `cms/37146/UBSAN/slc7_amd64_gcc11` context points to UBSAN results. Go there and then click on the `summary.html` page to see the UBSAN results

cmsbuild commented:
-1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cdca02/22856/summary.html
**COMMIT**: 81e7728c2b39c5a608e7949dc3ff80e79f14d0b7
**CMSSW**: CMSSW_12_3_UBSAN_X_2022-03-04-1100/slc7_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37146/22856/install.sh` to create a dev area with all the needed externals and cmssw changes.

smuzaffar commented:
ok, I have forced set the status for `ubsan` comparison check, but should add the comment for ubsan now

clacaputo commented:
If I check the output in `10004.0` step3 log, the runtime error is still there

```
/cvmfs/cms-ib.cern.ch/nweek-02722/slc7_amd64_gcc11/external/gcc/11.2.1-f478fee2760dbd22aaabb4e3a8fe1640/include/c++/11.2.1/bits/stl_vector.h:1046:34: runtime error: reference binding to null pointer of type 'value_type'
    #0 0x2abc3b477a03  (/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37146/22856/CMSSW_12_3_UBSAN_X_2022-03-04-1100/lib/slc7_amd64_gcc11/libRecoTrackerMkFitCore.so+0x730a03)
    #1 0x2abc3b49126b  (/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/37146/22856/CMSSW_12_3_UBSAN_X_2022-03-04-1100/lib/slc7_amd64_gcc11/libRecoTrackerMkFitCore.so+0x74a26b)
    #2 0x2abc3ad1aefc  (/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc11/cms/cmssw/CMSSW_12_3_UBSAN_X_2022-03-04-1100/lib/slc7_amd64_gcc11/libRecoTrackerMkFitCMS.so+0x50efc)
    #3 0x2abc3a82fd8e  (/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc11/cms/cmssw/CMSSW_12_3_UBSAN_X_2022-03-04-1100/lib/slc7_amd64_gcc11/pluginRecoTrackerMkFitPlugins.so+0x65cd8e)
```

mmusich commented:
> If I check the output in 10004.0 step3 log, the runtime error is still there

I observe the same locally:
```console
cmsrel CMSSW_12_3_UBSAN_X_2022-03-04-1100
cd CMSSW_12_3_UBSAN_X_2022-03-04-1100/src/
cmsenv
git cms-merge-topic 37146
scramv1 b -j 20
runTheMatrix.py -l 10004.0 -t 4 -j 8
more 10004.0_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log | grep libRecoTrackerMkFitCore | wc -l 
2
```

osschar commented:
Oh rats. How does one setup the build environment for this?
Reading gcc man page ... is USER_CXXFLAGS="-fsanitize=undefined -fsanitize=null" sufficient?

osschar committed:
[MkFit] Fix UBSAN error II. - skip processing of iterations with no seeds

Shortcut MkFitProducer to produce an empty track vector when input
seed vector is empty.

In MkBuilder::import_seeds() assert that the input vector is not
empty.

osschar commented:
There were also cases where MkFitProducer input seed vector was empty. The commit above bypasses all calls to mkFit and registers an empty track vector into Event.

This helped get to the error in gdb:
```
USER_CXXFLAGS="-fsanitize=undefined -fsanitize=null -g -O0" scram b -j16
gdb --args cmsRun ....
### wait for stuff to get loaded
b __ubsan::ScopedReport::~ScopedReport
### there are multiple functions in namespace __ubsan one can break on
```

cmsbuild commented:
Pull request #37146 was updated. @jpata, @cmsbuild, @clacaputo, @slava77 can you please check and sign again.

makortel commented:
@cmsbuild, please test

mmusich commented:
with this version of the code and the recipe at https://github.com/cms-sw/cmssw/pull/37146#issuecomment-1061736202 I get:

```console
$ more 10004.0_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log | grep libRecoTrackerMkFitCore | wc -l 
0
```
so we should be in the clear.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cdca02/22956/summary.html
**COMMIT**: 1d83b153da7f8dd800c412ed805dc143139d3bbd
**CMSSW**: CMSSW_12_3_X_2022-03-08-2300/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/37146/22956/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-cdca02/10004.0_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 27 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3806002
 - DQMHistoTests: Total failures: 872
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3805108
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 1 files compared)
 - Checked 6 log files, 0 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2022-03-08-2300+cdca02/48809/triggerResults): no differences found

clacaputo commented:
I'm sees some reco differences in wf `139.001` in `recoTracks_generalTracks__reRECO_obj`
For example
![all_OldVSNew_RunMinimumBias2021wf139p001c_recoTracks_generalTracks__reRECO_obj_p](https://user-images.githubusercontent.com/6500336/157409434-fc7d9eba-2635-4085-9f2b-533bf0e8e555.png)

Although it's infinitesimal, I wouldn't expect it considering the purely technical nature of this PR. Can be related to the wf being a MinimumBias wf?

slava77 commented on RecoTracker/MkFitCore/src/MkBuilder.cc:
```diff
@@ -800,7 +799,9 @@ namespace mkfit {
-          if (layer_plan_it.is_pickup_only() || theEndCand == 0)
+          dprintf("  Number of candidates to process: %d, nHits in layer: %d\n", theEndCand, layer_of_hits.nHits());
+
+          if (layer_plan_it.is_pickup_only() || theEndCand == 0 || layer_of_hits.nHits() == 0)
```
is the skip here on empty layer completely benign? Would we appropriately get the invalid/missing hits and material updates processed?

osschar committed:
[MkFit] Fix UBSAN error III. - undo  skipping of layers with no hits.

This led to physics changes in events with empty layers as tracks
passing through these layers were not assigned missing hits.

This change reverts the layer skipping and instead changes
MatriplexPackers to take object pointers instead of references to set
the base for vector gather indexing. This avoid dereferencing a
nullptr in case of empty layers.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-37146/28767

- This PR adds an extra 44KB to repository

cmsbuild commented:
Pull request #37146 was updated. @jpata, @cmsbuild, @clacaputo, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-cdca02/22994/summary.html
**COMMIT**: b548ebf4d8d5bc55dc06c3444718bd0d271588fc
**CMSSW**: CMSSW_12_3_X_2022-03-09-1100/slc7_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/37146/22994/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-cdca02/10004.0_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3806002
 - DQMHistoTests: Total failures: 15
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3805964
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 49 files compared)
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 210 log files, 45 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_3_X_2022-03-09-1100+cdca02/48844/triggerResults): no differences found

clacaputo commented:
+reconstruction

- technical PR, fix for UBSAN
- checked locally
```bash
cmsrel CMSSW_12_3_UBSAN_X_2022-03-04-1100
cd CMSSW_12_3_UBSAN_X_2022-03-04-1100/src/
cmsenv
git cms-merge-topic 37146
scramv1 b -j 20
runTheMatrix.py -l 10004.0 -t 4 -j 8
more 10004.0_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_SingleGammaPt10+2017+SingleGammaPt10_pythia8_GenSim+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log | grep libRecoTrackerMkFitCore | wc -l 
0
```
- no reco changes expected or observed after b548ebf4d8d5bc55dc06c3444718bd0d271588fc

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
+1

slava77 commented:
@cms-sw/core-l2 @perrotta @qliphy 
do you need a backport to 12_3_X?

makortel commented:
I suppose it would be useful (to minimize UB in 12_3).
